### PR TITLE
[#667][#502] test: fix flaky test QuorumTest

### PR DESCRIPTION
**Pull request by xianjingfeng**
### What changes were proposed in this pull request?

Fix flaky test QuorumTest#case7

### Why are the changes needed?

200ms is too short and  it is easy to timeout.
Fix: #667 and #502

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

No need.

### CODE REVIEW

**Comment by kaijchen** [comment]
> ### How was this patch tested?
> No need.

Can you test it in CI (repeat it for multiple times) to prove it's fixed?
Example: https://github.com/apache/ratis/pull/827

Command for repeating a test: `mvn test -Dtest=QuorumTest#case7`

**Comment by xianjingfeng** [comment]
> > ### How was this patch tested?
> > No need.
> 
> Can you test it in CI (repeat it for multiple times) to prove it's fixed? Example: [apache/ratis#827](https://github.com/apache/ratis/pull/827)
> 
> Command for running a single test: `mvn test -Dtest=QuorumTest#case7`

I don't know how to test it repeatedly. The env `ITERATIONS` doesn't work. What is its principle?

https://github.com/xianjingfeng/incubator-uniffle/blob/cc62f9a7721bd56debe9266079656b5693fc12dd/.github/workflows/test.yml#L30

**Comment by kaijchen** [comment]
> 100x QuorumTest#case7 before this PR: https://github.com/kaijchen/incubator-uniffle/actions/runs/4329005416

Seems it was good, maybe you should consider `mvn test -Dtest=QuorumTest`, to see if the tests are interferencing with each other.

**Comment by kaijchen** [comment]
> 100x QuorumTest#case7 before this PR: https://github.com/kaijchen/incubator-uniffle/actions/runs/4329005416

You can search for `[INFO]  T E S T S` in the log.

**Comment by xianjingfeng** [comment]
> Is this ready for merging?

Not yet.

File: integration-test/common/src/test/java/org/apache/uniffle/test/QuorumTest.java [hunk 1]
```diff
@@ -427,18 +431,25 @@ public void case3() throws Exception {
     } catch (Exception e) {
       assertTrue(e.getMessage().startsWith("Get shuffle result is failed"));
     }
+    shuffleServers.set(1, createServer(1, tmpDir));
+    shuffleServers.get(1).start();
```
Comment (xianjingfeng) [inline integration-test/common/src/test/java/org/apache/uniffle/test/QuorumTest.java:435]: It maybe the root cause of #483. @jerqi @zuston

File: integration-test/common/src/test/java/org/apache/uniffle/test/QuorumTest.java [hunk 2]
```diff
@@ -427,18 +431,25 @@ public void case3() throws Exception {
...
+
   }
 
   @Test
   public void case4() throws Exception {
     String testAppId = "case4";
+    /** We must wait until server1 finish start, because {@link #case3} will restart server1. */
+    await().timeout(10, TimeUnit.SECONDS).until(
+        () -> isPortAvailable(SHUFFLE_SERVER_PORT + 1));
+    Thread.sleep(500);
```
Comment (xianjingfeng) [inline integration-test/common/src/test/java/org/apache/uniffle/test/QuorumTest.java:452]: If i remove this line, this ut case will fail sometimes. i don't find the root cause.

**Comment by xianjingfeng** [comment]
> > 100x QuorumTest#case7 before this PR: https://github.com/kaijchen/incubator-uniffle/actions/runs/4329005416
> 
> Seems it was good, maybe you should consider `mvn test -Dtest=QuorumTest`, to see if the tests are interferencing with each other.

Done. https://github.com/xianjingfeng/incubator-uniffle/actions/runs/4460964902

File: integration-test/common/src/test/java/org/apache/uniffle/test/QuorumTest.java (hunk 2 above)
Comment (kaijchen) [reply integration-test/common/src/test/java/org/apache/uniffle/test/QuorumTest.java:452]: > If i remove this line, this ut case will fail sometimes. i don't find the root cause.
> https://github.com/xianjingfeng/incubator-uniffle/actions/runs/4460910585/jobs/7834366795

Seems it still fail sometimes.
https://github.com/kaijchen/incubator-uniffle/actions/runs/4464176186/jobs/7840089348

**Comment by kaijchen** [comment]
Thanks @xianjingfeng for working on this issue.

Seems success rate of QuorumTest is good without any changes:
https://github.com/kaijchen/incubator-uniffle/actions/runs/4464172169
I'm not sure why, but maybe it's interfering with other tests?

**Comment by xianjingfeng** [comment]
> Thanks @xianjingfeng for working on this issue.
> 
> Seems success rate of QuorumTest is good without any changes: https://github.com/kaijchen/incubator-uniffle/actions/runs/4464172169 I'm not sure why, but maybe it's interfering with other tests?

I guess it's caused by concurrent execution.

File: integration-test/common/src/test/java/org/apache/uniffle/test/QuorumTest.java (hunk 1 above)
Comment (xianjingfeng) [reply integration-test/common/src/test/java/org/apache/uniffle/test/QuorumTest.java:435]: I think we should not use `static` here.
https://github.com/apache/incubator-uniffle/blob/d60d675d38c833b99b012a1f4c726a012ce93463/integration-test/common/src/test/java/org/apache/uniffle/test/IntegrationTestBase.java#L60-L61

File: integration-test/common/src/test/java/org/apache/uniffle/test/QuorumTest.java (hunk 2 above)
Comment (xianjingfeng) [reply integration-test/common/src/test/java/org/apache/uniffle/test/QuorumTest.java]: I change it to sleep 1 second and run for 5 times. And it looks normal.
https://github.com/xianjingfeng/incubator-uniffle/actions/runs/4465348796
PS: some errors are caused by download dependency failures.

Comment (xianjingfeng) [reply integration-test/common/src/test/java/org/apache/uniffle/test/QuorumTest.java]: I think I have solved this problem. https://github.com/apache/incubator-uniffle/pull/673/commits/3020eb05d91c71736272029f521d102777520183

https://github.com/xianjingfeng/incubator-uniffle/actions/runs/4485510017

By the way, i found that this method was added by you, I think you should be able to explain the underlying principle to me. @zuston 
https://github.com/apache/incubator-uniffle/blob/5fd68764d6fc9e085c89f9ddabd62d99eea6163e/integration-test/common/src/test/java/org/apache/uniffle/test/MultiStorageFaultToleranceBase.java#L56

**Comment by xianjingfeng** [comment]
> I'm not sure about this fix, looking at the CI of last 10 commits on master. There are only 2 failures of "Unexpected flush process" in ShuffleFlushManagerTest.
> 
> ```
>  in org.apache.uniffle.server.ShuffleFlushManagerTest
> Error:  writeTest  Time elapsed: 51.105 s  <<< FAILURE!
> org.opentest4j.AssertionFailedError: Unexpected flush process
> ```

I think this problem has nothing to do with this PR. Because `ShuffleFlushManagerTest` is not in `integration-test` module, so i think they are independent of each other.

**Comment by kaijchen** [comment]
> I think this problem has nothing to do with this PR. Because `ShuffleFlushManagerTest` is not in `integration-test` module, so i think they are independent of each other.

Yes, I was trying to say, this QuorumTest is not very flaky. So I'm not sure increasing timeout will help.
Glad to see you have found the root cause. Cc @zuston PTAL.

**Comment by xianjingfeng** [comment]
> > I think this problem has nothing to do with this PR. Because `ShuffleFlushManagerTest` is not in `integration-test` module, so i think they are independent of each other.
> 
> Yes, I was trying to say, this QuorumTest is not very flaky. So I'm not sure increasing timeout will help. Glad to see you have found the root cause. Cc @zuston PTAL.

There is another question that I can't understand. In the original logic, why does server1 stop in case3, and case4 can still run normally?

And i found other probloms in `QuorumTest`
1. Sometimes `getShuffleResult` fail is because of `DEADLINE_EXCEEDED`, and not because server1 stop
https://github.com/apache/incubator-uniffle/blob/7cb0b000cdf7260db3ac77f9d1742b8496c24cb2/integration-test/common/src/test/java/org/apache/uniffle/test/QuorumTest.java#L421-L429

2. `shuffleServers` is a static variable and elements inside will be modified. And this will affect other tests.
https://github.com/apache/incubator-uniffle/blob/d60d675d38c833b99b012a1f4c726a012ce93463/integration-test/common/src/test/java/org/apache/uniffle/test/IntegrationTestBase.java#L60-L61

In general, modifying the timeout period will help expose the problems in advance.

**Comment by kaijchen** [comment]
Cc @frankliee who is the original author.

**Pull request closed without merging**